packages/runtime-core/src/apiLifecycle.ts

```diff
@@ -95,3 +96,32 @@ export const onErrorCaptured = (
 ) => {
   injectHook(LifecycleHooks.ERROR_CAPTURED, hook, target)
 }
+
+export function onServerPrefetch<
+  T extends () => Promise<any> = () => Promise<unknown>
+>(handler: T) {
+  const target = currentInstance
+  if (target) {
+    if (isInSSRComponentSetup) {
+      const type = target.type as ComponentOptions
+      let hook = type.serverPrefetch
+      if (hook) {
+        // Merge hook
+        type.serverPrefetch = () =>
+          Promise.all([handler(), (hook as Function)()])
```

Comment:
The `serverPrefetch` hooks call needs to bind `currentInstance.proxy`, the same as here https://github.com/vuejs/vue-next/blob/master/packages/server-renderer/src/render.ts#L96, otherwise it will break the expected behavior.

```suggestion
          Promise.all([handler.call(target.proxy), (hook as Function).call(target.proxy)])
```
This is necessary when the user uses it like the following:

```js
const Comp = {
  async serverPrefetch() {
    this.xxx
  }
}
```